### PR TITLE
Rework on Task::TalkNPC

### LINKED ISSUE #40: autotalkcont bug
On some servers the lag sometimes makes openkore disconnect when talking to npcs if the config key autotalkcont is set to '1'

### LINKED ISSUE #367: Problems when 'talk no' is used instead of 'talk cont'
Using 'sendTalkCancel' (talk no | talknpc n) when a 'sendTalkContinue' (talk cont | talknpc c) is expected causes openkore to get disconnected on bRO and on iRO, in RMS renewal test server it causes the player to the stuck in the cell but won't disconnect it.

Since 'sendTalkCancel' is sent everytime a 'talknpc' command fails I think it should be fixed ASAP.

I would love if someone could provide me with information regarding when the client actually sends  'sendTalkCancel' to the server and in which situations it can be sent without problems.

Edit 1: Since apparently all conversations in the real client end with a 'npc_talk_cancel' after a 'npc_talk_close' shouldn't we send it ALWAYS?

### PR DESCRIPTION
**Pull request by Henrybk**
Changes suggested in #362 

Changes so far:
1 - Made 'talk' command use `Task::TalkNPC`
2 - Mapchanges now clear `NPC` from `AI` queue
3 - Now `TalkNPC` tasks are auto created if a npc conversation starts and we don't have one
4 - Now `autoTalkCont` is handled by `Task::TalkNPC` ( fixes #40 )
5 - Now the next conversation step will only be sent if the server has responded to the last one, this was done to avoid DC's when the server is lagging, instead we only cancel the task.
6 - Changed `$talk{'buyOrSell'}` to `
$ai_v{'npc_talk'}{'talk'} eq
buy_or_sell`
7 - Removed `talk no` and `talknpc n` ( fixes #367 )
8 - `talk cancel` will always be sent after `talk close`
9 - Moved some dupped subs to `Receive.pm`
10 - Changed some `%talk` changes in `Send.pm`
11 - Rewrote all npc talk steps logic in `TalkNPC`
12 - Added a kind of intelligent logic to try to end a conversation after a bug happens
13 - Now you can add multiple steps at once using the command `talk`
14 - Now all talknpc sequences are validated before the conversation starts
15 - Messages like `Type 'talk xxxx' to continue conversation` won't appear if the next conversation step is defined.

TODO:
1 - Test on npcs with no talk or buy/sell interaction (private healer npcs that activate on click)

### CODE REVIEW

**Comment by Henrybk** [comment]
Also I think we should implement these ideas:
http://forums.openkore.com/viewtopic.php?f=36&t=182288
http://forums.openkore.com/viewtopic.php?f=36&t=10125

File: src/functions.pl [hunk 1]
```diff
@@ -644,7 +644,10 @@ sub initMapChangeVars {
 	undef %items;
 	undef %spells;
 	undef %incomingParty;
+	
 	undef %talk;
+	AI::clear("NPC");
```
Comment (farrainbow) [inline src/functions.pl]: TalkNPC task can hook map change event and mark itself as done (`setDone`) in the next `iterate` call.

File: src/Task/TalkNPC.pm [hunk 2]
```diff
@@ -70,33 +80,107 @@ sub new {
...
+			$self->{steps} = [parseArgs($self->{sequence})];
+		} else {
+			$self->{steps} = [];
+		}
+	} else {
+		if ($self->{type} eq 'talknpc') {
+			$self->{steps} = [parseArgs("x $self->{sequence}")];
+		} else {
+			push (@{$self->{steps}}, 'x');
+		}
```
Comment (farrainbow) [inline src/Task/TalkNPC.pm]: Maybe ignore `$self->{type}` and always use `$self->{sequence}` (looks like it's empty when `$self->{type} ne 'talknpc'`?).

File: src/functions.pl (hunk 1 above)
Comment (Henrybk) [inline src/functions.pl]: Because the task will only be cleaned in the first AI cycle after the mapchange, this can cause major bugs if another npc conversation starts before that, like a npc that starts conversation as soon as the character enters a certain map. (eg. Payon palace guards)

File: src/Task/TalkNPC.pm (hunk 2 above)
Comment (Henrybk) [inline src/Task/TalkNPC.pm]: Thats seems about right, I'll change it.

File: src/Task/TalkNPC.pm [hunk 3]
```diff
@@ -394,4 +626,9 @@ sub findTarget {
 	return undef;
 }
 
+sub nextStep {
+	my ($self, $step) = @_;
+	push(@{$self->{steps}}, $step);
+}
+
```
Comment (farrainbow) [inline src/Task/TalkNPC.pm]: Maybe name it `addSteps` and allow to pass many steps together in a single string argument, like when creating the task? Also it's not always would be "next" step since it just adds steps to the end of the sequence.

Comment (Henrybk) [inline src/Task/TalkNPC.pm]: I also noticed this earlier today, thanks for pointing it out.

File: src/Task/TalkNPC.pm [hunk 4]
```diff
@@ -70,33 +80,107 @@ sub new {
...
-	$self->{stage} = 'Not Started';
-	$self->{mapChanged} = 0;
+	$self->{stage} = NOT_STARTED;
+}
+
+sub find_and_set_target {
+	my ($self) = @_;
+	my $target = $self->findTarget($npcsList);
+	if ($target) {
+		debug "Target NPC " . $target->name() . " at ($target->{pos}{x},$target->{pos}{y}) found.\n", "ai_npcTalk";
```
Comment (farrainbow) [inline src/Task/TalkNPC.pm]: If `$target` is an `Actor`, you can just do `"Target $target at..."` to get a suitable string representation (with actor type, binID and possibly other info). This can be used in translated strings too: `TF("Found %s", $target)`.

File: src/Task/TalkNPC.pm [hunk 5]
```diff
@@ -70,33 +80,107 @@ sub new {
...
+	
+	if ($target) {
+		$self->{target} = $target;
+		$self->{ID} = $target->{ID};
+		lookAtPosition($self);
+	} elsif (exists $talk{nameID} && $ai_v{'npc_talk'}{'talk'} ne 'buy_or_sell') {#check if this is really necessary
+		$self->{ID} = $talk{ID};
+		$self->{target} = Actor::NPC->new;
+		$self->{target}->{appear_time} = time;
+		$self->{target}->{name} = 'Unknown';
```
Comment (farrainbow) [inline src/Task/TalkNPC.pm]: Which NPCs use this branch?

File: src/Task/TalkNPC.pm [hunk 6]
```diff
@@ -318,28 +444,110 @@ sub iterate {
...
+			# Usual end of a conversation
+			if (!%talk) {
+				$self->conversation_end;
+			
+			# No more steps but we still have a defined %talk
+			} else {
+				debug "After we finished talking with hte last npc another one started talking to us faster than we could predict.\n", "ai_npcTalk";
+				my $target = $self->find_and_set_target;
+				$self->{stage} = TALKING_TO_NPC;
+				$self->{time} = time;
```
Comment (farrainbow) [inline src/Task/TalkNPC.pm]: Which NPCs do this?

File: src/Task/TalkNPC.pm (hunk 2 above)
Comment (Henrybk) [inline src/Task/TalkNPC.pm]: Done

File: src/Network/Receive.pm [hunk 7]
```diff
@@ -3844,8 +3844,8 @@ sub npc_talk_close {
 		$messageSender->sendTalkCancel($ID);
 	}
 
-	$ai_v{npc_talk}{talk} = 'close';
-	$ai_v{npc_talk}{time} = time;
+	$ai_v{'npc_talk'}{'talk'} = 'close';
+	$ai_v{'npc_talk'}{'time'} = time;
```
Comment (farrainbow) [inline src/Network/Receive.pm]: Please next time do style changes in a separate commit, since it gets in the way of review in other places where style changes interleave with actual changes.

File: src/Task/TalkNPC.pm (hunk 5 above)
Comment (Henrybk) [inline src/Task/TalkNPC.pm]: NPCs scripts that start conversation without an actual npc 'body', like automatic bot checkers and automatic npc conversations on map changes.

File: src/Task/TalkNPC.pm (hunk 6 above)
Comment (Henrybk) [inline src/Task/TalkNPC.pm]: Any conversation that starts right after another one ends, 'payon 158 246 Guardsman' made this necessary before I added the deletion of the task at mapchange, don't know of any other.

File: src/Network/Receive.pm (hunk 7 above)
Comment (Henrybk) [inline src/Network/Receive.pm]: Sorry, did it without thinking :P

File: src/Network/Receive.pm [hunk 8]
```diff
@@ -3824,4 +3824,154 @@ sub hat_effect {
...
+
+sub npc_talk_continue {
+	my ($self, $args) = @_;
+	my $ID = substr($args->{RAW_MSG}, 2, 4);
+	my $name = getNPCName($ID);
+
+	$ai_v{'npc_talk'}{'talk'} = 'next';
+	$ai_v{'npc_talk'}{'time'} = time;
+	
+	message TF("%s: Type 'talk cont' to continue talking\n", $name), "npc";
```
Comment (farrainbow) [inline src/Network/Receive.pm]: Maybe put this message into TalkNPC task as well (and display it only when `autoTalkCont` is not set, and also when there are no more steps so user interaction is needed)?

Actually, maybe display all of those "type ... to continue talking" messages only when there are no more steps in the current sequence, so they don't appear at all when conversations are fully automated?

File: src/Task/TalkNPC.pm (hunk 4 above)
Comment (Henrybk) [inline src/Task/TalkNPC.pm]: Done

File: src/Network/Receive.pm (hunk 8 above)
Comment (Henrybk) [inline src/Network/Receive.pm]: Good idea

File: src/Task/TalkNPC.pm [hunk 9]
```diff
@@ -121,81 +211,83 @@ sub iterate {
...
+	# This is where things may bug in npcs which have no chat (private healers)
+	} elsif (!$ai_v{'npc_talk'}{'time'} && timeOut($self->{time}, $timeResponse)) {
 		# If NPC does not respond before timing out, then by default, it's
 		# a failure.
 		$messageSender->sendTalkCancel($self->{ID});
 		$self->setError(NPC_NO_RESPONSE, T("The NPC did not respond."));
 
-	} elsif (timeOut($ai_v{npc_talk}{time}, 0.25)) {
+	} elsif ($self->{stage} == TALKING_TO_NPC && timeOut($ai_v{'npc_talk'}{'time'}, 1.5)) {
 		# 0.25 seconds have passed since we last talked to the NPC.
```
Comment (farrainbow) [inline src/Task/TalkNPC.pm]: Comment became wrong...

File: src/Task/TalkNPC.pm [hunk 10]
```diff
@@ -206,103 +298,137 @@ sub iterate {
...
+			# Get the buy list in a shop.
+			} elsif ($step =~ /^b$/i) {
 				$messageSender->sendNPCBuySellList($talk{ID}, 0);
-			# Bulk buy solution.
-			} elsif ($step =~ /^b(\d+),(\d+)/i) {
+				
+			# Click the cancel button in a shop.
+			} elsif ($step =~ /^e$/i) {
+				$ai_v{'npc_talk'}{'talk'} = 'close';
+				$self->conversation_end;
```
Comment (farrainbow) [inline src/Task/TalkNPC.pm]: "e" shouldn't click the cancel button and it doesn't seem it's done there; so, wrong comment?

File: src/Task/TalkNPC.pm [hunk 11]
```diff
@@ -318,28 +444,110 @@ sub iterate {
...
+	# After a 'npc_talk_done' hook we must always send a 'npc_talk_cancel' after a timeout
+	} elsif ($self->{stage} == AFTER_NPC_CLOSE) {
+		return unless (timeOut($self->{time}, 1));
+		#Now 'n' step is totally unnecessary as we always send it but this must be done for backwards compatibility
+		if ( $self->{steps}[0] =~ /^n/i ) {
+			shift(@{$self->{steps}});
+		}
+		$self->{time} = time;
+		$self->{stage} = AFTER_NPC_CANCEL;
+		$messageSender->sendTalkCancel($self->{ID});
```
Comment (farrainbow) [inline src/Task/TalkNPC.pm]: I wonder if all clients do this.

Comment (Henrybk) [inline src/Task/TalkNPC.pm]: That needs confirmation, I tested on bRO and RMS renewal test server, it is up for changes.

File: src/Task/TalkNPC.pm (hunk 10 above)
Comment (Henrybk) [inline src/Task/TalkNPC.pm]: It's an old comment from the original code I haven't updated.

File: src/Task/TalkNPC.pm (hunk 9 above)
Comment (Henrybk) [inline src/Task/TalkNPC.pm]: Thanks, will fix.

File: src/Task/TalkNPC.pm (hunk 10 above)
Comment (Henrybk) [inline src/Task/TalkNPC.pm]: Actually in all servers I tested that is right, we don't send anything to the server when the client clicks the cancel button in a shop, so we just have to assume it's closed. (I ay be wrong, needs more testing)

File: src/Task/TalkNPC.pm (hunk 9 above)
Comment (Henrybk) [inline src/Task/TalkNPC.pm]: Done

File: src/Task/TalkNPC.pm (hunk 11 above)
Comment (Henrybk) [inline src/Task/TalkNPC.pm]: this code in receive.pm makes me think it is also important in other servers such as pRO and iRO.
`# I noticed that the RO client doesn't send a 'talk cancel' packet
	# when it receives a 'npc_talk_closed' packet from the server'.
	# But on pRO Thor (with Kapra password) this is required in order to
	# open the storage.
	#
	# UPDATE: not sending 'talk cancel' breaks autostorage on iRO.
	# This needs more investigation.
	#if (!$talk{canceled}) {
	#	$messageSender->sendTalkCancel($ID);
	#}`

File: src/Task/TalkNPC.pm [hunk 12]
```diff
@@ -351,20 +559,44 @@ sub target {
...
+		$ai_v{'npc_talk'}{'talk'} = 'close';
+	
+	} elsif ( $ai_v{'npc_talk'}{'talk'} eq 'store' ) {
+		$self->conversation_end;
+		$ai_v{'npc_talk'}{'talk'} = 'close';
+		
+	} elsif ( $ai_v{'npc_talk'}{'talk'} eq 'sell' ) {
+		$self->conversation_end;
+		$ai_v{'npc_talk'}{'talk'} = 'close';
+
```
Comment (farrainbow) [inline src/Task/TalkNPC.pm]: I'm not sure this is a good idea to have by default.

What happens if I have incomplete conversation sequence in talknpc command? Would I be able to finish it manually?

Which scenarios we want to accomplish with auto-finishing the conversation, if, for example, we already have `autoTalkCont` set?

We can have this but maybe there should be a config option to enable auto-finishing the conversation, and options to set text and number.

File: src/Task/TalkNPC.pm (hunk 10 above)
Comment (farrainbow) [inline src/Task/TalkNPC.pm]: Purpose of "e" is to just suspend the NPC conversation and allow other AI parts to do their stuff (buy/sell items, for example). So just the comment is wrong.

File: src/Task/TalkNPC.pm (hunk 11 above)
Comment (farrainbow) [inline src/Task/TalkNPC.pm]: So maybe we need to put that whole comment right there.

File: src/Task/TalkNPC.pm (hunk 12 above)
Comment (Henrybk) [inline src/Task/TalkNPC.pm]: This is only called when there's a bug in the sequence or the npc fails to respond correctly.

When the sequence is not complete it will only wait for a new step.

Comment (farrainbow) [inline src/Task/TalkNPC.pm]: Well, when we usually want to have that? Is this just for random NPCs who automatically open a conversation?

Comment (farrainbow) [inline src/Task/TalkNPC.pm]: Well, when we usually want to have that? I can't imagine useful scenarios of just ignoring faulty NPC sequences if they were purposely configured. Also NPC conversation would timeout after a while, wouldn't it?

Comment (farrainbow) [inline src/Task/TalkNPC.pm]: I'm fine with hiding it behind config option and further testing, though.

File: src/Task/TalkNPC.pm (hunk 11 above)
Comment (Henrybk) [inline src/Task/TalkNPC.pm]: I will move it.

File: src/Task/TalkNPC.pm (hunk 12 above)
Comment (Henrybk) [inline src/Task/TalkNPC.pm]: What will we do when the sequence fails though? Before it would send Talk cancel ans get openkore DC'ed, this code prevents that, should it just clear all remaining steps and next the task open for new steps?

File: src/Task/TalkNPC.pm (hunk 11 above)
Comment (Henrybk) [inline src/Task/TalkNPC.pm]: Done

File: src/Task/TalkNPC.pm (hunk 12 above)
Comment (farrainbow) [inline src/Task/TalkNPC.pm]: Well anything definitely would be improvement over DC'ing. Clearing the remaining steps sounds okay for me (after that either something would add more steps or NPC will timeout). We need input from more people on that.

**Comment by farrainbow** [comment]
>TODO:
>3 - Create a way to add multiple steps in a single command.

Just allow to do `->addSteps("c c r0")` and parse it the same way as task does when starting.

File: src/Task/TalkNPC.pm [hunk 13]
```diff
@@ -70,33 +80,119 @@ sub new {
...
+		$target = undef;
+	}
+	
+	if (exists $talk{nameID}) {
+		$self->{steps} = [parseArgs($self->{sequence})];
+	} else {
+		$self->{steps} = [parseArgs("x $self->{sequence}")];
+		undef $ai_v{'npc_talk'}{'time'};
+		undef $ai_v{'npc_talk'}{'talk'};
+	}
```
Comment (farrainbow) [inline src/Task/TalkNPC.pm]: What about this? (obv. addSteps needs to be able to handle this first)

```perl
	unless (exists $talk{nameID}) {
		$self->addSteps('x');
		undef $ai_v{'npc_talk'}{'time'};
		undef $ai_v{'npc_talk'}{'talk'};
	}
	$self->addSteps($self->{sequence});
```

File: src/Task/TalkNPC.pm (hunk 12 above)
Comment (farrainbow) [inline src/Task/TalkNPC.pm]: @itsrachelfish @allanon what do you think about this?

File: src/Task/TalkNPC.pm (hunk 13 above)
Comment (Henrybk) [inline src/Task/TalkNPC.pm]: Done

**Comment by Henrybk** [comment]
@farrainbow Feel free to make any changes you please, maintainers can edit this pull request :P

**Comment by farrainbow** [comment]
>TODO:
>2 - Add new ways to start a conversation (talk => binID, talknpc => coordinates, talkname => name, talkregex => regexOnName, etc)

A single command can handle all the cases.

>3 - Rework talk command to be able to give binId and steps in only one command like talknpc (eg. talk 0, resp 1, cont, text hello, resp 5)

Commas not really needed even.

I think commands should be changed separately and not with this PR - https://github.com/OpenKore/openkore/issues/390

File: src/Task/TalkNPC.pm (hunk 12 above)
Comment (Henrybk) [inline src/Task/TalkNPC.pm]: I added the config key and 4 ways to deal with the problem.

File: src/Task/TalkNPC.pm [hunk 14]
```diff
@@ -121,81 +221,93 @@ sub iterate {
...
-			if ($target) {
-				$self->{target} = $target;
-				$self->{ID} = $target->{ID};
-				$self->{steps} = [parseArgs("x $self->{sequence}")];
-				undef $ai_v{npc_talk}{time};
-				undef $ai_v{npc_talk}{talk};
-				lookAtPosition($self);
-			}
+			if ($target || %talk) {
+				unless (exists $talk{nameID}) {
```
Comment (farrainbow) [inline src/Task/TalkNPC.pm]: I guess `unless (%talk) {` would do the same thing there?

**Comment by Henrybk** [comment]
I found a bug that needs to be fixed, when using `Task::TalkNPC` to route to the npc and the character approaches and auto-talk npc and it auto-starts a conversation the task won't be able to end it.

**Comment by Henrybk** [comment]
Another bug: If Task::TalkNPC creates a subtask MapRoute, which also creates a subtasks TalkNPC (eg. to use kafra teleport), all subtasks hooks will be heard by the original Task::TalkNPC::handle_npc_talk.

**Comment by Henrybk** [comment]
3 More bugs:
1 - When using Task::TalkNPC to route to the npc and the character approaches and auto-talk npc and it auto-starts a conversation the task won't be able to end it.
2- If Task::TalkNPC creates a subtask MapRoute, which also creates a subtasks TalkNPC (eg. to use kafra teleport), all subtasks hooks will be heard by the original Task::TalkNPC::handle_npc_talk.
3 - Sometimes when using npcs to change maps (eg. kafra) 2 `npc_talk_done` will be sent by the server, bugging the conversation

**Comment by Henrybk** [comment]
The only big problems I still see are:
1 - Dealing with npcs with no interaction
2 - Dealing with unwanted conversation when there's no talknpc task
3 - Dealing with unwanted conversation when there's a talknpc task

**Comment by farrainbow** [comment]
I need to review this more, but I can't say I like the latest changes.

IMO in general this task (and all other tasks) should be able to have several independent instances.

Regarding approach argument: it should modify distFromGoal of MapRoute task instead, otherwise things may fail on strange maps.

Regarding approaching the NPC we want to talk to: since we can't command to approach "non-existing" NPC and there are always (map)+x+y arguments, we can just check whether an NPC which starts talking with us is the NPC we're already have a task to talk with (which didn't start yet - that can be done with creating a hook which NPC tasks would listen to and set a return value if it's "their" NPC), and only create a new task if it's a different one (and we encountered it while trying to go to another NPC). That could solve some of remaining bugs.

Checking for all subtasks to see something isn't a thing we like to do. It smells bad and TalkNPC task can be used in other tasks as well, and we don't really want to go find all instances of it. Also consider a setup when there's a Task::Chained with two Task::TalkNPC inside of it, or more complicated setups. It can help if we move adding hooks to task's activate method, or maybe even later, if possible (maybe not with the current code).

Also I've managed to do all pre-talking stuff (routing, waiting for movement to stop, finding target, initializing steps) task-way, resulting changes look simple, so later I can add that.

**Comment by Henrybk** [comment]
Oh, well, thanks. I didn't think the approaching code looked so bad, the other changes are a bit sly :P.

**Comment by farrainbow** [comment]
Maybe it would be easier to not do any routing inside of `Task::TalkNPC` and let `talknpc` command just create a `Task::Chained` with `Task::MapRoute` and `Task::TalkNPC` in the chain (if needed). If we then move hook addition to the `activate()`, then only one instance of `Task::TalkNPC` would catch them, even if on a way to some NPC we encounter another one. For NPCs which need approaching that would work too with some tweaks (the specified conversation steps would be activated and used only after routing is done). What do you think about this?

**Comment by farrainbow** [comment]
Example of what I mean:

---
**Task::Interaction** *(map, x, y, distance, task)* - route to the specified actor, find it, and do a *task*-based specified interaction with it

What it will do:
- make a `MapRoute` subtask
- listen to `actor_display` packet, compare with coordinates provided and save actor ID if we found the actor we need
- when routing is done, activate the provided in arguments *task* (that would be TalkNPC task for now, but we can make different ones later)
- listen to `npc_talk` packet, in case of "approaching NPC", and ONLY IF it's the same NPC ID as we've found previously, stop the routing subtask and activate the talking task the same way as in the previous point, but also provide it with the data of initial packet we received (just call some method); if NPC ID don't match, it should be ignored by this task and other things should be set up in such a way so new task would be created to handle the talking with that unwanted NPC (we can return values from hooks to do that easily)
---
If you like that and it handles all the cases we need, I can implement it.

The only problem I can see if we try to approach NPC which doesn't exist (so no info about its coordinates) but which still starts talking to us. Considering there may be various NPC like this, we can maybe add a *actorID* argument to specify ID of NPC we want to approach, if we really need to handle this case.

**Comment by farrainbow** [comment]
[18:07] <allanon> if we use coordinates, we need to ignore hidden npcs
[18:07] <allanon> iRO stacks the event and non-event npcs on the same cell

**Comment by farrainbow** [comment]
If needed: npc conversation step which would wait for NPC to start the talk first (for edge cases and lags with "approaching").

**Comment by farrainbow** [comment]
I'll create another branch for TalkNPC rework itself, and another one for dealing with routing.
